[Java Spring OAS3] Fix numerous OAS3 related Bugs

cachescrubber opened the pull request:
<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->

#9775 introduced in 5.3.1 causes numerous bugs when generating code.

This PR fixes the following Bugs

#11168 
#11165 
#11164 
#11163 
#11007
#11203 

This PR implements the following Requests / Improvements

#10409 

This PR resolves other, older PRs

 #10879

<!-- Please check the completed items below -->
### PR checklist
 
- [ x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x ] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x ] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

cachescrubber commented:
Cc: @wing328 , @welshm. Please have a look at this.

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:
```diff
@@ -8,11 +8,11 @@ package {{package}};
 import io.swagger.v3.oas.annotations.Parameter;
 import io.swagger.v3.oas.annotations.media.Content;
 import io.swagger.v3.oas.annotations.media.Schema;
-import io.swagger.v3.oas.annotations.parameters.RequestBody;
```
I think there's value in having both annotations present - but open to removing if no one else feels the same way

welshm replied:
This was my proposed fix

https://github.com/OpenAPITools/openapi-generator/pull/10766/files

cachescrubber replied:
I would address issue when we would actually need `io.swagger.v3.oas.annotations.parameters.RequestBody`. So far, the import is unused, hence the removal.

welshm replied:
It is used - by deleting this import the "@RequestBody" annotation is changing to

cachescrubber replied:
And this is how it should be - `org.springframework.web.bind.annotation.RequestBody` ist the annotation required technically by Spring Mvc or OpenFeign to handle a request. `io.swagger.v3.oas.annotations.parameters.RequestBody` is never used, as far as I can see.

If we need them both, they should be fully qualified

```
    @org.springframework.web.bind.annotation.RequestBody   // Spring
    @io.swagger.v3.oas.annotations.parameters.RequestBody  // Swagger
```
Question is : Do we actually need to support `@io.swagger.v3.oas.annotations.parameters.RequestBody`? What is it good for? Do Tools like SpringFox or SpringDoc support the annotation?

welshm replied:
I believe it is supported by SpringDoc

welshm replied:
Thoughts on hvaing both annotations?

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/libraries/spring-boot/pom.mustache:
```diff
@@ -24,7 +27,7 @@
         <groupId>org.springframework.boot</groupId>
         <artifactId>spring-boot-starter-parent</artifactId>
-        <version>{{#java8}}2.3.3.RELEASE{{/java8}}{{^java8}}1.5.12.RELEASE{{/java8}}</version>
+        <version>2.5.8</version>
```
Intentionally changing this? Seems to be java8 related as opposed to OAS3

cachescrubber replied:
The generated pom include outdated, unsupported dependencies. spring-boot 3.3 is EOL. I upgraded several outdated dependecies "en passant".

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/libraries/spring-boot/pom.mustache:
```diff
@@ -138,9 +116,12 @@
-        {{/useSpringfox}}
         {{#withXml}}
         <!-- XML processing: Jackson -->
+          <dependency>
```
note: Indentation looks off

cachescrubber replied:
fixed

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/libraries/spring-cloud/pom.mustache:
```diff
@@ -99,48 +105,39 @@
-{{/hateoas}}
-{{#useBeanValidation}}
+        {{/hateoas}}
+        {{#useBeanValidation}}
```
So maybe a general question - I assume by having this match the indentation of the pom file this is causing white space to be generated, which is likely why it was fully left aligned. 

Are there any community guidelines on this?

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/allowableValues.mustache:
```diff
@@ -0,0 +1 @@
+{{#allowableValues}}allowableValues = "{{#values}}{{{.}}}{{^-last}}, {{/-last}}{{#-last}}{{/-last}}{{/values}}"{{/allowableValues}}
```
Does this handle enums and containers properly? I don't see a change to "values.mustache" but maybe that's already correct

cachescrubber replied:
There is no values.mustache. Which file do you refer to?

welshm replied:
I was making an assumption about the reference to #values - does this work properly with enum values and containers then?

welshm reviewed (commented):
Thanks for fixing a lot of the issues I introduced! I think the paramDoc mustache is a great simplification overall and makes this much easier to understand

cachescrubber commented:
Re #11168. `@io.swagger.v3.oas.annotations.Hidden` is not applicable to Parameters. The correct replacement for
`@ApiIgnore` would be `@Parameter(hidden = true)` IMO.

welshm commented:
> Re #11168. `@io.swagger.v3.oas.annotations.Hidden` is not applicable to Parameters. The correct replacement for
> 
> `@ApiIgnore` would be `@Parameter(hidden = true)` IMO.

That seems reasonable to me

cachescrubber commented:
Short update before Xmas. I duplicated all test configs under `bin/configs/spring*.yaml` and added the **oas3** flag. Executing the scripts revealed even more problems, some of which I have addressed with the last commits. Basically WIP working through the list of test script configurations.

cachescrubber commented:
@welshm

Your OAS3 implementation basically disables springfox. At the time beeing, springfox moved on and supported for openapi3 is
available with the 3.0.0 release. Should we revert the following?

```
        // Springfox cannot be used with oas3 or apiFirst or reactive. So, write the property back after determining
        // whether it should be enabled or not.
        boolean useSpringFox = false;
        if (!apiFirst && !reactive && !oas3) {
            useSpringFox = true;
            additionalProperties.put("useSpringfox", true);
        }
        writePropertyBack("useSpringfox", useSpringFox);

```

whydoievenneedthis commented on modules/openapi-generator/src/main/resources/JavaSpring/pathParams.mustache:
```diff
@@ -1 +1 @@
-{{#isPathParam}}{{#useBeanValidation}}{{>beanValidationPathParams}}{{/useBeanValidation}}{{#oas3}}@Parameter(name = "{{baseName}}", description ={{/oas3}}{{^oas3}}@ApiParam(value ={{/oas3}} "{{{description}}}"{{#required}}, required = true{{/required}}{{#allowableValues}}, allowableValues = "{{#enumVars}}{{#lambdaRemoveDoubleQuote}}{{{value}}}{{/lambdaRemoveDoubleQuote}}{{^-last}}, {{/-last}}{{#- ...
+{{#isPathParam}}{{#useBeanValidation}}{{>beanValidationPathParams}}{{/useBeanValidation}} {{>paramDoc}} @PathVariable("{{baseName}}") {{>optionalDataType}} {{paramName}}{{/isPathParam}}
```
Any chance you could also push in here the `{{#isDate}}` and `{{#isDateTime}}` blocks from the `queryParams.mustache` below, as a side-thought?

cachescrubber replied:
See the last commit. I introduced `dateTimeParam.mustache` and call it from queryParams and pathParams. This is basically the same functionality as #10879 but with a new template to deduplicate the code.

Are there other param types where the DateTimeFormat annotation would make sense? headerParams? formParams?

Do you have any OpenApi specifications to include in the tests?

whydoievenneedthis replied:
Much appreciated. Minimal yaml for the path variable use case would be:

```yaml
openapi: 3.0.3
info:
  title: Api Documentation
  version: '1.0'
paths:
  /thingy/{date}:
    get:
      operationId: get
      parameters:
        - name: date
          in: path
          required: true
          schema:
            type: string
            format: date
            example: '2021-01-01'
      responses:
        '200':
          description: OK
```

I can imagine this to be useful both for form data and header, but I don't have a use case for either of those.

cachescrubber replied:
The last commit applies the @DateTimeFormat annotation consistently across the different parameter types. A test configuration is added based on your example spec.

welshm commented:
> @welshm
> 
> 
> 
> Your OAS3 implementation basically disables springfox. At the time beeing, springfox moved on and support for openapi3 is
> 
> available with the 3.0.0 release. Should we revert the following?
> 
> 
> 
> ```
> 
>         // Springfox cannot be used with oas3 or apiFirst or reactive. So, write the property back after determining
> 
>         // whether it should be enabled or not.
> 
>         boolean useSpringFox = false;
> 
>         if (!apiFirst && !reactive && !oas3) {
> 
>             useSpringFox = true;
> 
>             additionalProperties.put("useSpringfox", true);
> 
>         }
> 
>         writePropertyBack("useSpringfox", useSpringFox);
> 
> 
> 
> ```
> 
> 

Last I was reading about SpringFox, general consensus seemed to be for OAS 3 SpringDoc was greatly preferred. 

Re-enabling SpringFox support seems fine but I imagine could be done as a separate PR as I could see that likely needing a whole lot more changes

cachescrubber commented:
@welshm Agreed regarding Springfox. The last commit is +1y old and the current 3.0.0 starter is incompatible to spring-boot 2.6.x. Springfox is currently unmaintained.

oas=true and 5.3.1 + this PR will also resolve ##10409.

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/formParams.mustache:
```diff
@@ -1 +1,4 @@
-{{#isFormParam}}{{^isFile}}{{#oas3}}@Parameter(name = "{{{baseName}}}", description = {{/oas3}}{{^oas3}}@ApiParam(value = {{/oas3}}"{{{description}}}"{{#required}}, required = true{{/required}}{{#allowableValues}}, allowableValues = "{{#values}}{{{.}}}{{^-last}}, {{/-last}}{{#-last}}{{/-last}}{{/values}}"{{/allowableValues}}{{^isContainer}}{{#defaultValue}}, defaultValue = "{{{.}}}"{{/defaultValu ...
\ No newline at end of file
+{{#isFormParam}}
+  {{^isFile}}{{>paramDoc}}{{#useBeanValidation}} @Valid{{/useBeanValidation}} @RequestPart(value = "{{baseName}}"{{#required}}, required = true{{/required}}{{^required}}, required = false{{/required}}){{>dateTimeParam}} {{{dataType}}} {{paramName}}{{/isFile}}
```
note: Spacing/newlines here look odd

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/paramDoc.mustache:
```diff
@@ -0,0 +1 @@
+{{#oas3}}@Parameter(name = "{{{baseName}}}", description = "{{{description}}}"{{#required}}, required = true{{/required}}, schema = @Schema({{> allowableValues }}{{^isContainer}}{{#defaultValue}}, defaultValue = "{{{.}}}"{{/defaultValue}}{{/isContainer}})){{/oas3}}{{^oas3}}@ApiParam(value = "{{{description}}}"{{#required}}, required = true{{/required}}{{#allowableValues}}, {{> allowableValues }}{ ...
```
If there's no allowable values defined or defaults, does `@Schema()` work as an annotation?

Also, I this removes the `defaultValue` applied to the `@Parameter` annotation (one level up) - does that still work?

cachescrubber replied:
> If there's no allowable values defined or defaults, does @Schema() work as an annotation?

An empty `@Schema()` is the default anyway, Not pretty, but ok.

> Also, I this removes the defaultValue applied to the @Parameter annotation (one level up) - does that still work?

`@Parameter`  hast no defaultValue. IIRC. This one of the bugs introduced. `defaultValue` and `allowableValues`  moved to `@Schema`.

But I identified another problem. Maybee somone more experienced in mustache than I can help. When `allowableValues` is empty, but `defaultValue` is set. `schema = @Schema(, defaultValue = "11.2"))` is generated, which will obviously not compile. How to deal with the comma when there is nota single value of Schema which is always set?

whydoievenneedthis replied:
Move the `defaultValue` first. If it's set, good, if not, then make it `defaultValue = ""`, which is the default there. After that, if the `allowableValue` is set, it can safely have the comma before it.

cachescrubber replied:
I actually used `description=""` because defaultValue is not used at all times. It's usage is constrained by the `isContainer`  flag.

welshm reviewed (commented):
Looks like Bitrise failed to download a maven depedency.

CircleCI failed on OpenAPIDocumentationConfig

```
[ERROR] COMPILATION ERROR : 
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/spring-mvc-default-value/src/main/java/org/openapitools/configuration/OpenAPIDocumentationConfig.java:[14,48] cannot find symbol
  symbol:   class RelativePathProvider
  location: package springfox.documentation.spring.web.paths
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/spring-mvc-default-value/src/main/java/org/openapitools/configuration/OpenAPIDocumentationConfig.java:[49,53] cannot find symbol
  symbol:   class RelativePathProvider
  location: class org.openapitools.configuration.OpenAPIDocumentationConfig
```

Looks like that also needs to be updated for full Springfox removal

cachescrubber commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:
```diff
@@ -923,6 +924,17 @@ public void postProcessModelProperty(CodegenModel model, CodegenProperty propert
+    @Override
+    public CodegenModel fromModel(String name, Schema model) {
+        CodegenModel codegenModel = super.fromModel(name, model);
+        if (oas3) {
```
Actually this is how it is - the swagger2 imports are hard coded in `AbstractJavaCodegen` and can only be removed reliably in a post processor. They are not controlled by a mustache template. To change this, all Java based generators would have to be modified.

cachescrubber commented:
> 
> Looks like that also needs to be updated for full Springfox removal

Ok, git it. I have to revert to the latest 2.9.2 release of springfox. Since springfox is on removal, there is no value in upgrading all templates to 3.0.0 (which removes several deprected apis, including RelativePathProvider). A commit fixing this is coming soon.

cachescrubber commented:
@wing328

I introduced new test configs. Do I have to register them somewhere in order to be executed by the CI build?

- bin/configs/spring-boot-beanvalidation-no-nullable-oas3.yaml
- bin/configs/spring-boot-delegate-oas3.yaml
- bin/configs/spring-boot-implicitHeaders-oas3.yaml
- bin/configs/spring-boot-oas3.yaml
- bin/configs/spring-boot-reactive-oas3.yaml
- bin/configs/spring-boot-useoptional-oas3.yaml
- bin/configs/spring-cloud-async-oas3.yaml
- bin/configs/spring-cloud-date-time-oas3.yaml
- bin/configs/spring-cloud-oas3.yaml
- bin/configs/spring-cloud-petstore-feign-spring-pageable-oas3.yaml

cachescrubber commented:
@welshm Support for the implicitHeaders feature was also incomplete.

It missed to import ` io.swagger.v3.oas.annotations.Parameters`.  Swagger2 `@ApiImplicitParams` where generated in to the OAS3 `@Parameters`.

wing328 commented:
> I introduced new test configs. Do I have to register them somewhere in order to be executed by the CI build?

You can add them under https://github.com/OpenAPITools/openapi-generator/blob/master/pom.xml#L1174 but that may result in CI job fails to complete as the job may run over 60 minutes.

If that's the case, I'll move it to other CI later.

cachescrubber commented:
One more thing...

In OAS3 `@Schema`, allowableValues is now a String[] array:

` String[] allowableValues() default {};`

In Swagger v2 `@ApiParam` it used to be a specially formatted String

`String defaultValue() default "";`

another commit will take care of this.

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/bodyParams.mustache:
```diff
@@ -1 +1,2 @@
-{{#isBodyParam}}{{#oas3}}@Parameter(name ={{/oas3}}{{^oas3}}@ApiParam(value ={{/oas3}} "{{{description}}}"{{#required}}, required = true{{/required}} {{^isContainer}}{{#defaultValue}}, defaultValue = "{{{.}}}"{{/defaultValue}}{{/isContainer}}{{#allowableValues}}, {{#oas3}}schema = @Schema({{/oas3}}allowableValues = "{{{.}}}"{{#oas3}}){{/oas3}}{{/allowableValues}})  {{#useBeanValidation}} @Valid{{ ...
\ No newline at end of file
+{{#isBodyParam}}{{>paramDoc}}{{#useBeanValidation}} @Valid{{/useBeanValidation}} @RequestBody{{^required}}(required = false){{/required}} {{^reactive}}{{{dataType}}}{{/reactive}}{{#reactive}}{{^isArray}}Mono<{{{dataType}}}>{{/isArray}}{{#isArray}}Flux<{{{baseType}}}>{{/isArray}}{{/reactive}} {{paramName}}
+{{/isBodyParam}}
```
ideally this is probably on the previous line?

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/cookieParams.mustache:
```diff
@@ -1 +1,2 @@
-{{#isCookieParam}}{{#useBeanValidation}}{{>beanValidationQueryParams}}{{/useBeanValidation}}{{#oas3}}@Parameter(name = "{{{baseName}}}", description = {{/oas3}}{{^oas3}}@ApiParam(value = {{/oas3}}"{{{description}}}"{{#required}}, required = true{{/required}}{{#allowableValues}}, allowableValues = "{{#enumVars}}{{#lambdaEscapeDoubleQuote}}{{{value}}}{{/lambdaEscapeDoubleQuote}}{{^-last}}, {{/-last ...
\ No newline at end of file
+{{#isCookieParam}}{{#useBeanValidation}}{{>beanValidationQueryParams}}{{/useBeanValidation}}{{>paramDoc}} @CookieValue("{{baseName}}"){{>dateTimeParam}} {{>optionalDataType}} {{paramName}}
+{{/isCookieParam}}
```
Same about previous line

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/pathParams.mustache:
```diff
@@ -1 +1,2 @@
-{{#isPathParam}}{{#useBeanValidation}}{{>beanValidationPathParams}}{{/useBeanValidation}}{{#oas3}}@Parameter(name = "{{baseName}}", description ={{/oas3}}{{^oas3}}@ApiParam(value ={{/oas3}} "{{{description}}}"{{#required}}, required = true{{/required}}{{#allowableValues}}, allowableValues = "{{#enumVars}}{{#lambdaRemoveDoubleQuote}}{{{value}}}{{/lambdaRemoveDoubleQuote}}{{^-last}}, {{/-last}}{{#- ...
+{{#isPathParam}}{{#useBeanValidation}}{{>beanValidationPathParams}}{{/useBeanValidation}}{{>paramDoc}} @PathVariable("{{baseName}}"){{>dateTimeParam}} {{>optionalDataType}} {{paramName}}
+{{/isPathParam}}
```
Previous line?

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/queryParams.mustache:
```diff
@@ -1 +1,2 @@
-{{#isQueryParam}}{{#useBeanValidation}}{{>beanValidationQueryParams}}{{/useBeanValidation}}{{#oas3}}@Parameter(name = "{{{baseName}}}", description = {{/oas3}}{{^oas3}}@ApiParam(value = {{/oas3}}"{{{description}}}"{{#required}}, required = true{{/required}}{{#allowableValues}}, {{#oas3}}schema = @Schema({{/oas3}}allowableValues = "{{#values}}{{{.}}}{{^-last}}, {{/-last}}{{#-last}}{{/-last}}{{/val ...
\ No newline at end of file
+{{#isQueryParam}}{{#useBeanValidation}}{{>beanValidationQueryParams}}{{/useBeanValidation}}{{>paramDoc}}{{#useBeanValidation}} @Valid{{/useBeanValidation}}{{^isModel}} @RequestParam(value = {{#isMap}}""{{/isMap}}{{^isMap}}"{{baseName}}"{{/isMap}}{{#required}}, required = true{{/required}}{{^required}}, required = false{{/required}}{{^isContainer}}{{#defaultValue}}, defaultValue = "{{{.}}}"{{/defa ...
+{{/isQueryParam}}
```
Previous line

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/paramDoc.mustache:
```diff
@@ -0,0 +1 @@
+{{#oas3}}@Parameter(name = "{{{baseName}}}", description = "{{{description}}}"{{#required}}, required = true{{/required}}, schema = @Schema(description = ""{{#allowableValues}}, {{> allowableValues }}{{/allowableValues}}{{^isContainer}}{{#defaultValue}}, defaultValue = "{{{.}}}"{{/defaultValue}}{{/isContainer}})){{/oas3}}{{^oas3}}@ApiParam(value = "{{{description}}}"{{#required}}, required = true ...
```
style: spacing around `{{> allowableValues }}` is inconsistent with other files (e.g. `{{>dateTimeParam}}`

cachescrubber replied:
I used the spacing to match the style elsewhere in the project. Just search for the exact string.

welshm reviewed (approved):
I'm happy to follow up with a separate change for the annotation for `@RequestBody` to have both.

Otherwise this is looking pretty good to me

cachescrubber commented:
> I'm happy to follow up with a separate change for the annotation for `@RequestBody` to have both.
Ok with me. Anyway, I would suggest to check the [SpringDoc](https://springdoc.org) documentation if there is a use case where this annotation is actually used.

cachescrubber commented:
@wing328, @welshm 

I think the OAS3 story is now complete. I won't commit anything else from my side other then as a response to your feedback.

Please advise me how to proceed. 

PS: Happy new year to you.

welshm commented:
> @wing328, @welshm 
> 
> 
> 
> I think the OAS3 story is now complete. I won't commit anything else from my side other then as a response to your feedback.
> 
> 
> 
> Please advise me how to proceed. 
> 
> 
> 
> PS: Happy new year to you.

Happy New year!

Need to wait for @wing328 to review and merge it at this point I believe

welshm commented on modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:
```diff
@@ -153,13 +156,18 @@ public interface {{classname}} {
         consumes = { {{#consumes}}"{{{mediaType}}}"{{^-last}}, {{/-last}}{{/consumes}} }{{/hasConsumes}}{{/singleContentTypes}}
     )
-    {{#jdk8-default-interface}}default {{/jdk8-default-interface}}{{#responseWrapper}}{{.}}<{{/responseWrapper}}ResponseEntity<{{>returnTypes}}>{{#responseWrapper}}>{{/responseWrapper}} {{#delegate-method}}_{{/delegate-method}}{{operationId}}({{#allParams}}{{>queryParams}}{{>pathParams}}{{>headerParams}}{{>bodyParams}}{{>formParams}}{{>cookieParams}}{{^-last}},{{/-last}}{{#-last}}{{#reactive}}, { ...
+    {{#jdk8-default-interface}}default {{/jdk8-default-interface}}{{#responseWrapper}}{{.}}<{{/responseWrapper}}ResponseEntity<{{>returnTypes}}>{{#responseWrapper}}>{{/responseWrapper}} {{#delegate-method}}_{{/delegate-method}}{{operationId}}(
```
Hmmm looks like `@Hidden` may still be relevant from SpringDoc


```

Additionally, to @Hidden annotation from swagger-annotations, its possible to restrict the generated OpenAPI description using package or path configuration.

```

As well as the guide on migrating from SpringFox

```
@ApiIgnore → @Parameter(hidden = true) or @Operation(hidden = true) or @Hidden

```

Although for method annotation rather than parameter

cachescrubber replied:
Maybe. We should add `@Hidden` in a separat PR when it's needed.

welshm commented:
May also fix 

https://github.com/OpenAPITools/openapi-generator/issues/11203

wing328 commented:
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/OpenAPITools/openapi-generator/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

cachescrubber commented:
Ups, I used $COMPANY email address ....

@wing328 I could squash commit the whole branch in a new one and submit another PR. Would that be acceptable?

wing328 commented:
@cachescrubber you can simply add your company email address as a secondary email address in your Github account. Let me know if you need help on that.

cachescrubber commented:
@wing328 Ok, I used the secondary email approach. Commits are now linked to my account.

wing328 commented:
Tested the new samples locally and all compiles without issues.

sabrina981 commented:
Downs Brandon your account is closed for non payment with insurance bill
code for your account information

On Sun, 2 Jan 2565 BE at 2:49 AM cachescrubber ***@***.***>
wrote:

> Ups, I used $COMPANY email address ....
>
> @wing328 <https://github.com/wing328> I could squash commit the whole
> branch in a new one and submit another PR. Would that be acceptable?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenAPITools/openapi-generator/pull/11181#issuecomment-1003682625>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUVZYNCQLLZH2IVVLVEY2O3UUAGS7ANCNFSM5KUX5ANA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

cemo commented:
@wing328 when will the release be cut?  Is it possible to cut a patch release?
